packages/plugins/response-cache/src/plugin.ts

```diff
@@ -320,90 +327,45 @@ export function useResponseCache<PluginContext extends Record<string, any> = {}>
-                    }
-                    if (
-                      fieldName === '__leave' &&
-                      /**
-                       * The visitResult function is called for each field in the selection set.
-                       * But visitResult function looks for __typename field visitor even if it is not there in the document
-                       * So it calls __typename field visitor twice if it is also in the selection set.
-                       * That's why we need to count the number of times it is called.
-                       *
-                       * Default call of __typename https://github.com/ardatan/graphql-tools/blob/master/packages/utils/src/visitResult.ts#L277
-                       * Call for the field node https://github.com/ardatan/graphql-tools/blob/master/packages/utils/src/visitResult.ts#L272
-                       */ typenameCalled < 2
-                    ) {
-                      return (root: any) => {
-                        delete root.__typename;
-                        return root;
-                      };
-                    }
-
-                    if (
-                      scopePerSchemaCoordinate[`${typename}.${fieldName}`] === 'PRIVATE' &&
-                      !sessionId
-                    ) {
-                      skip = true;
-                      return;
-                    }
-
-                    if (idFields.includes(fieldName)) {
-                      return (id: string) => {
-                        identifier.set(`${typename}:${id}`, { typename, id });
-                        return id;
-                      };
-                    }
-
-                    return undefined;
-                  },
-                });
-              },
-            },
-          ),
-        );
+      function processResult(data: any) {
+        if (data == null || typeof data !== 'object') {
+          return;
+        }
+        if (Array.isArray(data)) {
+          for (const item of data) {
+            processResult(item);
+          }
+          return;
+        }
+        const typename = data.__responseCacheTypeName;
+        delete data.__responseCacheTypeName;
+        if (!skip) {
+          if (
+            ignoredTypesMap.has(typename) ||
+            (scopePerSchemaCoordinate[typename] === 'PRIVATE' && !sessionId)
+          ) {
+            skip = true;
+            return;
+          }
+          types.add(typename);
+          if (typename in ttlPerType) {
+            currentTtl = calculateTtl(ttlPerType[typename], currentTtl);
+          }
+          for (const fieldName in data) {
+            if (scopePerSchemaCoordinate[`${typename}.${fieldName}`] === 'PRIVATE' && !sessionId) {
+              skip = true;
+            }
+            if (!skip) {
+              const id = data[fieldName];
+              if (idFields.includes(fieldName)) {
+                identifier.set(`${typename}:${id}`, { typename, id });
+              }
```

Comment:
We probably can move the id lookup in the condition ?
```suggestion
              if (idFields.includes(fieldName)) {
                const id = data[fieldName];
                identifier.set(`${typename}:${id}`, { typename, id });
              }
```